titan-compiler/checker.lua

```diff
@@ -241,17 +241,32 @@ function checkstat(node, st, errors)
         st:add_symbol(node.name, node)
         node._type = node._type or typefromnode(node.type, errors)
     elseif tag == "Stat_Decl" then
-        if node.decl.type then
-          checkstat(node.decl, st, errors)
-          checkexp(node.exp, st, errors, node.decl._type)
-        else
-          checkexp(node.exp, st, errors)
-          node.decl._type = node.exp._type
-          checkstat(node.decl, st, errors)
+        for i = 1, #node.decl do
+            local decl_i = node.decl[i]
+            local exp_i = node.exp[i]
+            if exp_i then
+                if decl_i.type then
+                    checkstat(decl_i, st, errors)
+                    checkexp(exp_i, st, errors, decl_i._type)
+                else
+                    checkexp(exp_i, st, errors)
+                    decl_i._type = exp_i._type
+                    checkstat(decl_i, st, errors)
+                end
+                exp_i = trycoerce(exp_i, decl_i._type, errors)
+                node.exp[i] = exp_i
+                checkmatch("declaration of local variable " .. decl_i.name,
+                    decl_i._type, exp_i._type, errors, decl_i._pos)
```

Comment:
Couldn't we move those 4 lines to the then case?
If the type come from the expression, we don't need to coerce and verify.